src/services/StreamPeriodsService.ts

```diff
@@ -0,0 +1,190 @@
+
+	// Timestamp when stream period, virtual period end and end date filter stop overlapping
+	const streamPeriodEndTimestamp = Math.min(virtualPeriodEndTimestamp, endTimestamp);
+
+	const virtualStreamPeriod: VirtualStreamPeriod = {
+		startTime: streamPeriodStartTimestamp,
+		endTime: streamPeriodEndTimestamp,
+		amount: getAmountInTimespan(streamPeriodStartTimestamp, streamPeriodEndTimestamp, flowRate).toString(),
+		amountFiat: calculateVirtualStreamPeriodPrice(
+			streamPeriodStartTimestamp,
+			streamPeriodEndTimestamp,
+			flowRate,
+			priceData,
+		).toString(),
+	};
+
+	if (endTimestamp <= virtualPeriodEndTimestamp) return [virtualStreamPeriod];
+
+	const nextPeriodStartDate = fromUnixTime(virtualPeriodEndTimestamp + 1);
+	return [
+		virtualStreamPeriod,
+		...virtualizeStreamPeriod(streamPeriod, nextPeriodStartDate, endDate, period, priceData),
+	];
+}
+
+function getAmountInTimespan(startTimestamp: number, endTimestamp: number, flowRate: string): Decimal {
+	return new Decimal(flowRate).mul(new Decimal(endTimestamp - startTimestamp));
+}
+
+function getUniqueNetworkTokenAddresses(streamPeriods: StreamPeriodResult[]): NetworkToken[] {
+	return Object.values(
+		streamPeriods.reduce((tokens, streamPeriod) => {
+			const {
+				chainId,
+				token: { id, underlyingAddress },
+			} = streamPeriod;
+
+			// TODO: If no underlyingAddress then it is native super token?
+			return {
+				...tokens,
+				[`${chainId}-${underlyingAddress}`]: { chainId, token: underlyingAddress || id },
+			};
+		}, {}),
+	);
+}
+
+function getPeriodRelevantPriceData(startTimestamp: number, endTimestamp: number, priceData: TimespanPrice[]) {
+	const priceWhenPeriodStarts = priceData
+		.filter((timespanPrice) => timespanPrice.start <= startTimestamp)
+		.reduce((startTimespanPrice: TimespanPrice | undefined, timespanPrice: TimespanPrice) => {
+			if (!startTimespanPrice || startTimespanPrice.start < timespanPrice.start) return timespanPrice;
+			return startTimespanPrice;
+		}, undefined);
+
+	const priceDataDuringTimePeriod = priceData.filter(
+		(timespanPrice) => timespanPrice.start > startTimestamp && timespanPrice.start <= endTimestamp,
+	);
+
+	return priceWhenPeriodStarts ? [priceWhenPeriodStarts, ...priceDataDuringTimePeriod] : priceDataDuringTimePeriod;
+}
+
+function calculateVirtualStreamPeriodPrice(
+	startTimestamp: number,
+	endTimestamp: number,
+	flowRate: string,
+	priceData: TimespanPrice[],
+) {
+	const relevantPriceData = getPeriodRelevantPriceData(startTimestamp, endTimestamp, priceData);
+
+	return mapPriceDataToVirtualStreamPeriodRecursive(
+		new Decimal(0),
+		new Decimal(flowRate),
+		startTimestamp,
+		endTimestamp,
+		relevantPriceData,
+	);
+}
+
+function mapPriceDataToVirtualStreamPeriodRecursive(
+	currentTotal: Decimal,
+	flowRate: Decimal,
+	startTimestamp: number,
+	endTimestamp: number,
+	priceData: TimespanPrice[],
+) {
+	const [timespanPrice, ...remainingPriceData] = priceData;

```

Comment:
we're assuming it's correctly ordered, right?

Reply:
Yup ;)